osu.Game/Storyboards/StoryboardSprite.cs

```diff
@@ -118,94 +115,43 @@ public StoryboardSprite(string path, Anchor origin, Vector2 initialPosition)
-            generateCommands(generated, getCommands(g => g.Scale, triggeredGroups), (d, value) => d.Scale = new Vector2(value), (d, value, duration, easing) => d.ScaleTo(value, duration, easing));
-            generateCommands(generated, getCommands(g => g.Rotation, triggeredGroups), (d, value) => d.Rotation = value, (d, value, duration, easing) => d.RotateTo(value, duration, easing));
-            generateCommands(generated, getCommands(g => g.Colour, triggeredGroups), (d, value) => d.Colour = value, (d, value, duration, easing) => d.FadeColour(value, duration, easing));
-            generateCommands(generated, getCommands(g => g.Alpha, triggeredGroups), (d, value) => d.Alpha = value, (d, value, duration, easing) => d.FadeTo(value, duration, easing));
-            generateCommands(generated, getCommands(g => g.BlendingParameters, triggeredGroups), (d, value) => d.Blending = value, (d, value, duration, _) => d.TransformBlendingMode(value, duration),
-                false);
-
-            if (drawable is IVectorScalable vectorScalable)
-            {
-                generateCommands(generated, getCommands(g => g.VectorScale, triggeredGroups), (_, value) => vectorScalable.VectorScale = value,
-                    (_, value, duration, easing) => vectorScalable.VectorScaleTo(value, duration, easing));
-            }
+            HashSet<string> appliedProperties = new HashSet<string>();
 
-            if (drawable is IFlippable flippable)
-            {
-                generateCommands(generated, getCommands(g => g.FlipH, triggeredGroups), (_, value) => flippable.FlipH = value, (_, value, duration, _) => flippable.TransformFlipH(value, duration),
-                    false);
-                generateCommands(generated, getCommands(g => g.FlipV, triggeredGroups), (_, value) => flippable.FlipV = value, (_, value, duration, _) => flippable.TransformFlipV(value, duration),
-                    false);
-            }
+            // For performance reasons, we need to apply the commands in chronological order.
+            // Not doing so will cause many functions to be interleaved, resulting in O(n^2) complexity.
+            IEnumerable<IStoryboardCommand> commands = Commands.AllCommands;
+            commands = commands.Concat(loopingGroups.SelectMany(l => l.AllCommands));
 
-            foreach (var command in generated.OrderBy(g => g.StartTime))
-                command.ApplyTo(drawable);
-        }
-
-        private void generateCommands<T>(List<IGeneratedCommand> resultList, IEnumerable<CommandTimeline<T>.TypedCommand> commands,
-                                         DrawablePropertyInitializer<T> initializeProperty, DrawableTransformer<T> transform, bool alwaysInitialize = true)
-        {
-            bool initialized = false;
-
-            foreach (var command in commands)
+            foreach (var command in commands.OrderBy(c => c.StartTime))
```

Comment:
This looks... expensive...

- `AllCommands`, which is a `SelectManyIterator`
- which is then linq-concatenated with another `SelectManyIterator`
- and then finally linq-sorted

Was linq overhead measured here? It looks like it would be massive. I guess this is only ran during load which makes it less of a killer, but surely this can be improved by materialising somewhere?

Reply:
I have been going quite easy on the code because it's a once-off, yeah. I will revise it.